### PR TITLE
Explicit unpack buffers

### PR DESCRIPTION
**Pull request by abadams**
In master, an input buffer by the name of foo places all sorts of symbols in scope, some of them redundant, and it's not clear which ones a load needs. foo.host and foo are used interchangeably.

In this PR, only foo.buffer is placed in scope automatically by an Argument or embedded Buffer of name foo. The rest of the unpacking is done explicitly in IR. The host symbol is unpacked to "foo". Loads and Stores require a symbol with the matching name (of type Handle()) to be in scope. This means you can load and store to arbitrary Handles. Allocate nodes are just RAII let nodes now. They call malloc (by default) and place a symbol with the given name in scope.

Defining precisely what "foo" means removes many uses of address_of. If you want the beginning of an allocation it's just Variable::make(Handle(), "foo");

Calls and Provides are at a higher level of abstraction, and require an enclosing Realize node or buffer Argument with the same name, rather than just a symbol.

### CODE REVIEW

File: src/HexagonOffload.cpp [hunk 1]
```diff
@@ -67,18 +67,19 @@ class InjectHexagonRpc : public IRMutator {
...
 
     Expr state_var(const std::string& name, Type type) {
-        Expr& var = state_vars[name];
-        if (!var.defined()) {
-            auto storage = Buffer<void *>::make_scalar(name + "_buf");
-            storage() = nullptr;
-            var = Load::make(type_of<void*>(), storage.name(), 0, storage, Parameter(), const_true());
-        }
-        return var;
+        return Let::make(name, state_var_ptr(name, type),
```
Comment (abadams) [inline src/HexagonOffload.cpp]: Here is an example of loading directly from a symbol of type Handle, instead of requiring an Allocate node in scope.

**Comment by ronlieb** [comment]
looking forward to this fixing the hexagon issue.
Thx

**Comment by steven-johnson** [comment]
LGTM

**Comment by ronlieb** [comment]
Thanks Andrew, the PR fixed all but one problem instance

 

Internal error at /local/mnt2/ronl/Halide-upstream/Halide/src/Scope.h:111 triggered by user code at /local/mnt2/ronl/Halide-upstream/Halide/test/correctness/simd_op_check.cpp:212:

Symbol 'in_u8' not found

 

…

 

cd bin/build/tmp ; /local/mnt2/ronl/Halide-upstream/Halide/bin/correctness_simd_

op_check

valign(v*,v*,#7)

vlalign(v*,v*,#7)

valign(v*,v*,r*)

valign(v*,v*,r*)

valign(v*,v*,#6)

vlalign(v*,v*,#6)

valign(v*,v*,r*)

valign(v*,v*,r*)

vunpack(v*.ub)

vunpack(v*.ub)

vunpack(v*.uh)

vunpack(v*.uh)

vunpack(v*.b)

vunpack(v*.b)

vunpack(v*.h)

vunpack(v*.h)

vunpack(v*.ub)

vunpack(v*.ub)

vunpack(v*.b)

vunpack(v*.b)

vadd(v*.b,v*.b)

Internal error at /local/mnt2/ronl/Halide-upstream/Halide/src/Scope.h:111 triggered by user code at /local/mnt2/ronl/Halide-upstream/Halide/test/correctness/simd_op_check.cpp:212:

Symbol 'in_u8' not found

bash: line 1:  4463 Aborted                 (core dumped) /local/mnt2/ronl/Halide-upstream/Halide/bin/correctness_simd_op_check

 

---

Qualcomm Innovation Center, Inc. is a member of Code Aurora Forum, hosted by The Linux Foundation

**Comment by abadams** [comment]
I'll take a look.

**Comment by abadams** [comment]
The host pointer is now provided by an explicit LetStmt that extracts the
host field of the input buffer instead of being magically in scope.
HexagonOptimize is stripping dead lets as a side-effect, using
expr_uses_var. But expr_uses_var does not consider a Load (or a Store) as a
use of the variable with the corresponding name. I'll just change
expr_uses_var to make loads and stores count as name uses.

**Pull request merged**